pkg/vpcmodel/grouping.go

```diff
@@ -0,0 +1,215 @@
+	res := []EndpointElem{}
+	subnetNameToNodes := map[string][]Node{} // map from subnet name to its nodes from the input
+	groupedSubnets := map[string]bool{}      // map from subnet name to bool indicating if it was grouped
+	for _, elem := range elemsList {
+		n, ok := elem.(Node)
+		if !ok {
+			continue
+		}
+		subnet := c.getSubnetOfNode(n)
+		subnetName := subnet.Name()
+		if _, ok := subnetNameToNodes[subnetName]; !ok {
+			subnetNameToNodes[subnetName] = []Node{}
+			groupedSubnets[subnetName] = false
+		}
+		subnetNameToNodes[subnetName] = append(subnetNameToNodes[subnetName], n)
+		if len(subnetNameToNodes[subnetName]) == len(subnet.Nodes()) {
+			res = append(res, subnet)
+			groupedSubnets[subnetName] = true
+		}
+	}
+	for subnetStr, isGrouped := range groupedSubnets {
+		if !isGrouped {
+			for _, node := range subnetNameToNodes[subnetStr] {
+				res = append(res, node)
+			}
+		}
+	}
+	return res
+}
+
+func (g *GroupConnLines) groupExternalAddresses() {
+	// phase1: group public internet ranges
+	res := []*GroupedConnLine{}
+	for src, nodeConns := range g.v.AllowedConnsCombined {
+		for dst, conns := range nodeConns {
+			if conns.IsEmpty() {
+				continue
+			}
+			connString := ConnStr(conns.String())
+			switch {
+			case dst.IsPublicInternet():
+				g.srcToDst.addPublicConnectivity(src, connString, dst)
+			case src.IsPublicInternet():
+				g.DstToSrc.addPublicConnectivity(dst, connString, src)
+			default:
+				res = append(res, &GroupedConnLine{src, dst, connString})
+			}
+		}
+	}
+	// add to res lines from  srcToDst and DstToSrc groupings
+	res = append(res, g.srcToDst.getGroupedConnLines(true)...)
+	res = append(res, g.DstToSrc.getGroupedConnLines(false)...)
+	g.GroupedLines = res
+}
+
+// assuming the  g.groupedLines was already initialized by previous step groupExternalAddresses()
+func (g *GroupConnLines) groupSubnetsSrcOrDst(srcGrouping bool) {
+	res := []*GroupedConnLine{}
+	// build map from str(dst+conn) to []src => create lines accordingly
+	groupingSrcOrDst := map[string][]*GroupedConnLine{}
+
+	// populate map groupingSrcOrDst
+	for _, line := range g.GroupedLines {
+		srcOrDst, dstOrSrc := line.getSrcOrDst(srcGrouping), line.getSrcOrDst(!srcGrouping)
+		if _, ok := srcOrDst.(Node); !ok {
+			res = append(res, line)
+			continue
+		}
+		key := dstOrSrc.Name() + ";" + string(line.Conn)
+		if _, ok := groupingSrcOrDst[key]; !ok {
+			groupingSrcOrDst[key] = []*GroupedConnLine{}
+		}
+		groupingSrcOrDst[key] = append(groupingSrcOrDst[key], line)
+	}
+
+	// update g.groupedLines based on groupingSrcOrDst
+	for _, linesGroup := range groupingSrcOrDst {
+		// if linesGroup.Src contains an entire subnet => group to one line with subnet
+		// else, keep separated lines
+		srcOrDstGroup := make([]EndpointElem, len(linesGroup))
+		for i, line := range linesGroup {
+			srcOrDstGroup[i] = line.getSrcOrDst(srcGrouping)
+		}
+		groupedSrcOrDst := subnetGrouping(srcOrDstGroup, g.c)
+		if groupedSrcOrDst != nil {
```

Comment:
Doesn't seem as if `groupedSrcOrDst` is ever going to be `nil`. Is this condition redundant?

Reply:
right, removed that condition (remained from older version)